lang/src/org/partiql/lang/eval/builtins/Time.kt

```diff
@@ -0,0 +1,116 @@
+package org.partiql.lang.eval.builtins
+
+import com.amazon.ion.IonStruct
+import com.amazon.ion.IonSystem
+import org.partiql.lang.errors.ErrorCode
+import org.partiql.lang.eval.err
+import org.partiql.lang.util.getOffsetHHmm
+import org.partiql.lang.util.propertyValueMapOf
+import org.partiql.lang.util.times
+import java.math.BigDecimal
+import java.math.RoundingMode
+import java.time.LocalTime
+import java.time.OffsetTime
+import java.time.ZoneOffset
+import java.time.format.DateTimeFormatter
+import kotlin.math.min
+
+/**
+ * Wrapper class representing the run time instance of TIME in PartiQL.
+ * - `TIME [(p)] HH:MM:ss[.ddd][+|-HH:MM]` PartiQL statement creates a run-time instance of this class with [zoneOffset] as [null].
+ * - `TIME WITH TIME ZONE [(p)] HH:MM:ss[.ddd][+|-HH:MM]` PartiQL statement creates a run-time instance of this class with both
+ * [localTime] and [zoneOffset] defined.
+ *
+ * @param localTime Represents the local time.
+ * @param zoneOffset Represents the time zone of the TIME instance.
+ *  If the [zoneOffset] is null, the zone offset is not defined for the instance.
+ * @param precision Represents the number of significant digits in fractional part of the second's value of the TIME instance.
+ *  The default precision is 9 meaning, the default precision is of nanoseconds.
+ *  If the precision is specified by the user, the fractional part of the second will be rounded to the specified precision.
+ *  For eg, `TIME (3) 23:46:58.1267` will be stored in this instance with [localTime] as `23:46:58.127000000` along with preserving the
+ *  [precision] value as 3.
+ *  Note that the [LocalTime] always stores the fractional part of the second in nanoseconds.
+ *  It is up to the application developers to make use of the preserved [precision] value as need be.
+ */
+data class Time(val localTime: LocalTime, val zoneOffset: ZoneOffset? = null, val precision: Int = MAX_PRECISION_FOR_TIME) {
+
+    init {
+        // Validate that the precision value is non-negative.
+        if (precision < 0) {
+            err(
+                message = "Specified precision for TIME should be a non-negative integer between 0 and 9 inclusive",
+                errorCode = ErrorCode.EVALUATOR_INVALID_PRECISION_FOR_TIME,
+                errorContext = propertyValueMapOf(),
+                internal = false)
+        }
+    }
+
+    companion object {
+
+        const val PARTIQL_TIME_ANNOTATION = "\$partiql_time"
+        const val HOURS_PER_DAY = 24
+        const val MINUTES_PER_HOUR = 60
+        const val SECONDS_PER_MINUTE = 60
+        const val SECONDS_PER_HOUR = SECONDS_PER_MINUTE * MINUTES_PER_HOUR
+        const val NANOS_PER_SECOND = 1000000000.0
+        const val MAX_PRECISION_FOR_TIME = 9
```

Comment:
It may seem logical for organizational purposes to put the functions and constants we want to be "static" in a companion object, however, I would strongly discourage that and instead use top-level declarations, that are `private` (ideally), or `internal` (if needed), and only `public` if intended to be accessed by customer code.

Aside from being idiomatic to Kotlin (when in Kotlin-land, do as the Kotlonians do :) ), using top-level declarations for these is preferable because `companion object` are *not* "static" at all, they actually needlessly create an anonymously typed singleton `object` instance with these fields.  Companion objects also pollute the API as seen by Java and seem to provide much in the way of actual value.

In general, I'd avoid the use of `companion object`s unless truly needed.  I haven't yet seen a circumstance where I felt using the `companion object` was *truly* required.  The alternatives are almost always preferable.

Some of the code in this library was written before we came to this conclusion about `companion object`s, so unfortunately you'll see a number that don't actually need to be there.

Reply:
I see. Thanks @dlurton for sharing the great insight. I have moved these to top level.